vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SnapshotManager.java

```
@@ -0,0 +1,370 @@
+    }
+    
+    /**
+     * 
+     * Download given space and related modules into exhibitionFolderPath
+     * @param space
+     * @param exhibitionFolderPath
+     * @param context
+     */
+    @Override
+    public void downloadSpace(Space space, String exhibitionFolderName,  SequenceHistory sequenceHistory) {
+
+        String spaceFolderPath = exhibitionFolderName + File.separator + space.getId();
+        storageEngineDownloads.createFolder(spaceFolderPath);
+
+        storeTemplateForSpace(space.getId(), spaceFolderPath , sequenceHistory);
+
+        String imagesFolderPath = spaceFolderPath + File.separator  + IMAGES_FOLDER_NAME;
+        storageEngineDownloads.createFolder(imagesFolderPath); 
+
+        // Copies the space image
+        storageManager.copyImageUploadsToDownloads(space.getImage(), imagesFolderPath) ;
+
+        List<IModuleLink> moduleLinks = space.getModuleLinks();
+
+        moduleLinks.forEach(moduleLink -> {
+
+            IModule module =   moduleLink.getModule();
+            downloadModule(module, space,  imagesFolderPath, spaceFolderPath);
+
+        });        
+    }
+    
+
+
+
+    
+    
+
+    /**
+     * Stores the processed template for space into spaceFolderPath.
+     * 
+     * @param directory
+     * @param spaceFolderPath
+     * @param context
+     */
+    @Override
+    public void storeTemplateForSpace(String directory, String spaceFolderPath,  SequenceHistory sequenceHistory) {
+        try {           
+            Context thymeleafContext = new Context();
+            populateContextForSpace(thymeleafContext, directory, sequenceHistory);
+            // add attributes to context
+            String response = springTemplateEngine.process("exhibition/downloads/spaceDownloadTemplate", thymeleafContext);
+            byte[] fileContent = response.getBytes();
+            storageEngineDownloads.storeFile(fileContent, directory+".html",spaceFolderPath );
+
+        } catch ( FileStorageException e) {
```

hm, if the file can't be stored then likely the whole process is flawed and the snapshot creation won't work properly. It's probably better to throw an exception in this case and abort.